File: frontend/src/playlist/RankingOption.js
import React, {useEffect, useRef} from 'react';
import PropTypes from "prop-types";

const RankingOption = (props) => {

    const audioRef = useRef(null);

    useEffect(() => {
        if (audioRef.current) {
            audioRef.current.volume = props.previewVolume;
        }
    }, [props.previewVolume]);

    const audioRefCallback = (e) => {
        if (e) {
            audioRef.current = e;
            e.volume = props.previewVolume;
        }
    };

    return (
        <div className={"ranking-option" + (props.optionsClickable ? " clickable" : "")}
             onClick={() => props.selectOption(props.optionData.track_uri)}>
            <div>
                <img className={"ranking-image"} src={props.imageMap.get(props.optionData.album_uri)}
                     alt={"Option 1 cover"}/>
            </div>
            <span
                className={"ranking-option-row hor-centered-text ranking-option-track-name"}>{props.optionData.track_name}</span>
            <div className={"hor-centered vert-centered ranking-option-row"}>
                <img className={"ranking-artist-image"} src={props.imageMap.get(props.optionData.artist_uri)}
                     alt={"Option 1 artist"}/>
                <span className={"ranking-artist-name"}>{props.optionData.artist_name}</span>
            </div>
            <span
                className={"ranking-option-row hor-centered-text ranking-option-album-name"}>{props.optionData.album_name}</span>
            <div className={"ranking-option-row audio-preview"}>
                {props.optionData.audio_preview_url ?
                    <audio src={props.optionData.audio_preview_url} controls={true} autoPlay={false}
                           ref={audioRefCallback}
                           onVolumeChange={(event) => {
                               props.setPreviewVolume(event.target.volume)
                           }}/> : <></>}
            </div>
            {!props.optionsClickable ? <div className={"grey-overlay"}/> : <></>}
        </div>
    );
};

RankingOption.props = {
    optionsClickable: PropTypes.any,
    selectOption: PropTypes.any,
    imageMap: PropTypes.any,
    optionData: PropTypes.any,
    previewVolume: PropTypes.any,
    setPreviewVolume: PropTypes.any
}

export default RankingOption;

File: backend/data_objects/RankingOption.py
import util


class RankingOption:
    def __init__(self, album_uri: str, album_name: str, album_image: bytes, track_uri: str, track_name: str,
                 artist_uri: str, artist_name: str, artist_image: bytes, audio_preview_url: str):
        self.album_uri = album_uri
        self.album_name = album_name
        self.album_image = album_image

        self.track_uri = track_uri
        self.track_name = track_name
        self.audio_preview_url = audio_preview_url

        self.artist_uri = artist_uri
        self.artist_name = artist_name
        self.artist_image = artist_image

    def to_dict(self):
        return {
            "album_uri": self.album_uri,
            "album_name": self.album_name,
            "album_image_b64": util.image_bytes_to_b64(self.album_image),
            "track_uri": self.track_uri,
            "track_name": self.track_name,
            "audio_preview_url": self.audio_preview_url,
            "artist_uri": self.artist_uri,
            "artist_name": self.artist_name,
            "artist_image_b64": util.image_bytes_to_b64(self.artist_image),
        }
